### PR TITLE
Support tree-shakeable icons

### PR DESCRIPTION
**Pull request by mshwery**
This PR builds on the work from @Rowno and @Chrischuck to get tree shake-able icons shipped in Evergreen.

Note: this is not quite ready for shipping, but it's close. We could probably ship after adding types for the new icons.

- [x] generate icon names export
- [x] generate individual icon files
- [x] generate map of icon names to icon components (for backwards compatibility in `<Icon icon='foo' />`)
- [x] generate index.js exports of all individual icons
- [x] remove `@blueprintjs/icons` from the final builds (only use generated types to reduce build size)
- [x] reduce build size from 350kb to ~240kb (closer to today's 212kb)
- [x] generate typedefs for individual components
- [x] remove usage of `<Icon />` internally wherever possible in a non-breaking way
- [x] update docs
- [ ] ship as a minor version!

|BEFORE (350kb gzipped, blueprintjs is duplicated)|AFTER (238kb gzipped,  no blueprintjs duplication)|
|---|---|
|Screen Shot 2020-04-15 at 11 01 20 AM|Screen Shot 2020-04-15 at 9 33 52 PM|

Screen Shot 2020-04-15 at 9 41 33 PM

### CODE REVIEW

**Comment by mshwery** [comment]
@Chrischuck had to split into my own branch since I don't have push permissions for your fork.

**Comment by mshwery** [comment]
Ok, I think this might be ready for prime time review!

I'm happy to remove the generated icons from git, but it might be nice to see when there are changes to the icons in future versions.

File: src/theme/src/default-theme/theme-helpers/index.js [hunk 1]
```diff
@@ -1,7 +1,14 @@
+import React from 'react'
 import { Intent } from '../../../../constants'
 import themedProperty from '../utils/themedProperty'
 import { colors, elevations } from '../foundational-styles'
 import { fontFamilies, headings, paragraph, text } from '../typography'
+import {
```
Comment (akleiner2) [inline src/theme/src/default-theme/theme-helpers/index.js:6]: ❤️

File: tools/generate-icons.js [hunk 2]
```diff
@@ -133,10 +133,7 @@ export const ${iconName} = memo(forwardRef((props, ref) => (
   // =====================
 
   const iconsExport = `/* Start generated icons */
-    export {
-      IconNames,\n
-      ${iconNames.join(',\n  ')}
-    } from './icons'
+    export * from './icons'
```
Comment (akleiner2) [inline tools/generate-icons.js:136]: Nice 👍

**Comment by mshwery** [comment]
Amazing foundation by @Chrischuck and @Rowno 👏

**Pull request merged**